File: src/communication/TCPServer.cpp
//
// Created by ga78cat on 14.03.2021.
//

#include <comm/communication/TCPServer.h>
#include <cassert>
#include <comm/socket/utils.h>

using namespace comm;

TCPServer::TCPServer(int port, int backlog) : port(port), backlog(backlog), socket(nullptr) {
    this->socketAcceptTimeout = {.tv_sec = SOCKET_ACCEPT_TIMEOUT_SECONDS, .tv_usec = 0};
    this->initServerSocket();
}

TCPServer::~TCPServer() {
    this->cleanup();
}

void TCPServer::cleanup() {
    delete this->socket;
    this->socket = nullptr;
}

Communication *TCPServer::acceptCommunication() {
    static fd_set readSockets;
    FD_ZERO(&readSockets);
    FD_SET(this->socket->getSocket(), &readSockets);

    int selectResult = select((int) this->socket->getSocket() + 1, &readSockets, nullptr, nullptr,
                              &(this->socketAcceptTimeout));
    if (selectResult > 0) {
        // The accept() call actually accepts an incoming connection
        auto *comm = new Communication();
        this->socket->accept(comm->getSocket(SocketType::TCP));
        return comm;
    } else if (selectResult == SOCKET_ERROR) {
        printLastError();
        this->initServerSocket();
        return nullptr;
    } else {
        // Timeout occurred!
        assert(selectResult == 0);
        return nullptr;
    }
}

void TCPServer::initServerSocket() {
    if (this->socket != nullptr) {
        this->socket->cleanup();
        this->socket->initialize(SocketType::TCP, nullptr, this->port);
    } else {
        this->socket = new Socket(SocketType::TCP, nullptr, this->port);
    }
    this->listen();
}

void TCPServer::listen() const {
    ::listen(this->socket->getSocket(), this->backlog);
}

File: csharp/communication/TCPServer.cs
using Comm.socket;
using System;
using System.Diagnostics;

namespace Comm.communication {
    class TCPServer {
        public TCPServer(int port, int backlog = 5) {
            this.port = port;
            this.backlog = backlog;
            this.socket = null;
            this.initServerSocket();
        }

        ~TCPServer() {
            Console.WriteLine("In TCPServer destructor!");
            this.cleanup();
        }

        public void cleanup() {
            this.socket.cleanup();
        }

        private void listen() {
            this.socket.getSocket().Listen(this.backlog);
        } 

        public Communication acceptCommunication() {
            try {
                System.Net.Sockets.Socket acceptedSocket = this.socket.getSocket().Accept();
                Debug.Assert(acceptedSocket.Blocking == this.socket.getSocket().Blocking);
                acceptedSocket.Blocking = true;
                // The accept() call actually accepts an incoming connection
                Communication comm = new Communication();
                Socket tcpSocket = null;
                this.socket.accept(ref tcpSocket, acceptedSocket);
                comm.setSocket(SocketType.TCP, socket);
                return comm;
            } catch (System.Net.Sockets.SocketException se) {
                int errorCode = se.ErrorCode;
                if (errorCode == 10035) {
                    // Timeout occurred!
                    return null; 
                } else {
                    Console.WriteLine("ErrorCode in acceptCommunication = " + errorCode);
                    this.initServerSocket();
                    return null;
                }
            }
        }

        private void initServerSocket() {
            if (this.socket != null) {
                this.socket.cleanup();
                this.socket.initialize(SocketType.TCP, null, this.port);
            } else {
                this.socket = new Socket(SocketType.TCP, null, this.port);
            }
            this.socket.getSocket().Blocking = false;
            this.listen();
        }

        private int port, backlog;
        private Socket socket;
    };
}
